src/binding/c/pt2pt_api.txt

```diff
@@ -260,8 +285,13 @@ MPI_Isend:
 {
     MPIR_Request *request_ptr = NULL;
 
-    mpi_errno = MPID_Isend(buf, count, datatype, dest, tag, comm_ptr,
-                           MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);
+    if (comm_ptr->remote_size == 1 && comm_ptr->comm_kind == MPIR_COMM_KIND__INTRACOMM) {
+        mpi_errno = MPID_Self_isend(buf, count, datatype, dest, tag, comm_ptr,
+                                    MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);
+    } else {
+        mpi_errno = MPID_Isend(buf, count, datatype, dest, tag, comm_ptr,
+                               MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);
+    }
```

Comment:
Conceptually, I agree these branches can live here. Practically, if we move them into ch4, then we *should* be able to save the request object bloat by adding the self struct to the MPID_REQUEST_DEV_DECL union. ch3 can probably live without this functionality.

Reply:
I'll move it inside ch4 then. It will be slightly simpler actually. But later we have to treat ch3 and ch4 differently in the upper layer for `MPI_Session_init`. I guess we can simply add a config option, e.g. `enable-local-session-init` and set it automatically depend on whether ch3 or ch4 is configured.

Reply:
I was assuming we'd just return an error or abort is a process attempted `MPI_Session_init` with ch3 like we do with partitioned communication.